### ISSUE TITLE
Error while doing "LastGoodCheckDB": Microsoft.Data.SqlClient.SqlException (0x80131904): When converting a varchar data type to a datetime data type, the value is out of range

### ISSUE THREAD

**Issue by SasaJ7** [trimble-oss/dba-dash#613]
Hello there!

We get an attached datetime convert error while doing "LastGoodCheckDB" on on our servers. You can find german error message stack to this post.

Error message:
Microsoft.Data.SqlClient.SqlException (0x80131904): Bei der Konvertierung eines varchar-Datentyps in einen datetime-Datentyp liegt der Wert außerhalb des gültigen Bereichs. Die Anweisung wurde beendet. 
(Translated: When converting a varchar data type to a datetime data type, the value is out of range. The statement has ended. )

Windows and SQL Server language: german 
Server Collation: Latin1_General_CI_AS
Number of Databases: 15 (same number of error messages)

Please help me to troubleshoot this message further and provide more info to fix this bug.

Thank you.

**Comment by DavidWiseman** [comment]
Hi,

This is the query used for collection:

https://github.com/trimble-oss/dba-dash/blob/main/DBADash/SQL/SQLLastGoodCheckDB.sql

If it's a older SQL version it gets its data from DBCC DBINFO and newer versions will use DATABASEPROPERTYEX.  In both cases it's casting the value to a DATETIME.

It might be worth running the query above on your server to troubleshoot.  You could try removing the CAST to see what the value is.

Let me know what you find.  It will be good to know the value it's trying to cast and if it's from DBCC DBINFO or from DATABASEPROPERTYEX.

Thanks,

David

**Comment by SasaJ7** [comment]
Hi David,

thanks to point in right direction.

I have multiple SQL Servers running with Language: english and german. 
This issue was shown on fully patched SQL Server 2014 with german language, but I found it on other Versions running on german language.

As you correctly pointed out, error pops on all selects containing cast function, where I concentrated to line 38 of your above script.

after troubleshooting it on english SQL 2016 and german SQL 2014, by implementing "SET DATEFORMAT mdy" fixes issues on your script.

I think you have similar issue on all other sql servers with languages different from english. 

How shall we procede on this case?

**Comment by DavidWiseman** [comment]
Thank you for the info.  The issue can be changed by using CONVERT instead of CAST for this line of code:

```sql
SELECT TOP(1) @DBID, CAST(Value as DATETIME)
```

To:

```sql
SELECT TOP(1) @DBID, CONVERT(DATETIME,Value,120)
```

I can get this fixed for the next release.

This issue will only impact older versions of SQL server that have a different date format.  The CAST is OK for newer versions of SQL server as `DATABASEPROPERTYEX(name,'LastGoodCheckDbTime') ` returns a SQL_VARIANT and the underlying type is DATETIME.  It's only on older versions where DBCC DBINFO is used and it needs to convert from a string.

What's interesting is I can't reproduce this on SQL 2014 - `DATABASEPROPERTYEX(name,'LastGoodCheckDbTime') ` is available on SQL 2014 (At least on SP3-CU4) so it shouldn't need to call DBCC DBINFO per database.  

Would you be able to check your version with this:
```sql
SELECT @@VERSION
```
I'm wondering if you might be on SQL 2012 (where I can reproduce the issue) or an older patch version of SQL 2014.

Thanks,

David

**Comment by SasaJ7** [comment]
Hi David,

my bad, it was one of unpatched SQL 2014 Servers.

Here's Version:
Microsoft SQL Server 2014 - 12.0.2000.8 (X64)   Feb 20 2014 20:04:26   Copyright (c) Microsoft Corporation  Standard Edition (64-bit) on Windows NT 6.3 <X64> (Build 17763: ) (Hypervisor) 

I should push for upgrade of that server.

Thanks for help with my issue!

Sasa

**Comment by DavidWiseman** [comment]
It looks like `DATABASEPROPERTYEX([DatabaseName], 'lastgoodcheckdbtime') was added to SQL 2014 in [SP3](https://support.microsoft.com/en-us/topic/kb4022619-sql-server-2014-service-pack-3-release-information-64f341d3-2b74-0d86-e857-b9fd0775e493) and also added to SQL 2016 in [SP2](https://support.microsoft.com/en-us/topic/kb4052908-sql-server-2016-service-pack-2-release-information-69d59b78-5632-5878-634f-58b16186aa0d).  

It's still a bug in DBA Dash as it should support older versions from SQL 2005 onwards and not be sensitive to language & culture settings used.

**Comment by DavidWiseman** [comment]
Fix to be included in the next release.

**Comment by SasaJ7** [comment]
David, thanks a lot! :)

**Comment by DavidWiseman** [comment]
Fixed in 2.40.2 🎉
